packages/grpc-js/src/server-call.ts

```diff
@@ -422,17 +423,24 @@ export class Http2ServerCallStream<
        * expect to be able to send an error over it. */
     });
 
+    this.stream.once('aborted', () => {
+      this.aborted = true;
+    });
+
     this.stream.once('close', () => {
       trace(
         'Request to method ' +
           this.handler?.path +
           ' stream closed with rstCode ' +
           this.stream.rstCode
       );
-      this.cancelled = true;
-      this.emit('cancelled', 'cancelled');
-      this.emit('streamEnd', false);
-      this.sendStatus({code: Status.CANCELLED, details: 'Cancelled by client', metadata: new Metadata()});
+
+      if (this.aborted) {
```

Comment:
I don't see the reason to have this. At a high level, the purpose of the `cancelled` event is to notify the server that it no longer needs to work on this operation because the client no longer needs the result. That is just as true if the TCP connection disappeared as if the client explicitly sent a cancellation.

Reply:
`close` event will fire on stream regardless of whether it was cancelled or gracefully closed. performance benefits of this is especially noticeable on unary requests - we are avoiding emitting multiple events and triggering associated events for respectful listeners.

for instance, one may have an abort controller listen to 'cancelled' events in case of server or client initiated stream cancels to schedule cleanup events, but the way it stands right now - regardless of how the request has finished the `cancelled` event is always fired and needless work would be triggered - as it stands right now `cancelled` is renamed `close` event, which isn't surfaced

so ideally cancelled would have it's own semantics to notify the user about a cancelled (meaning unfinished) stream and you could have `close` being emitted alongside of it to have the old semantics

Reply:
This is using `nice-grpc` wrapper around `@grpc/grpc-js`:

<img width="665" alt="Screen Shot 2022-10-17 at 15 42 41" src="https://user-images.githubusercontent.com/1713617/196297259-b241bfea-5447-410e-b049-bd65f1d5d1c3.png">

This is how things look like with "selective" `cancelled` event:

<img width="1727" alt="Screen Shot 2022-10-17 at 15 47 11" src="https://user-images.githubusercontent.com/1713617/196297765-aba8fbb5-cbc8-41e1-9008-f1ee1e8c6020.png">

```
ghz --insecure --proto ./src/hello.proto --call helloworld.Greeter.SayHello -d '{"who":"nodeconf"}' -n 300000 --connections 1 --count-errors --timeout 0 0.0.0.0:8080

Summary:
  Count:	300000
  Total:	10.66 s
  Slowest:	22.14 ms
  Fastest:	0.12 ms
  Average:	1.43 ms
  Requests/sec:	28151.58

Response time histogram:
  0.125  [1]      |
  2.326  [271660] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  4.527  [25860]  |∎∎∎∎
  6.728  [1801]   |
  8.929  [311]    |
  11.131 [82]     |
  13.332 [135]    |
  15.533 [16]     |
  17.734 [65]     |
  19.935 [19]     |
  22.136 [50]     |

Latency distribution:
  10 % in 0.85 ms
  25 % in 1.08 ms
  50 % in 1.26 ms
  75 % in 1.52 ms
  90 % in 2.28 ms
  95 % in 2.66 ms
  99 % in 4.20 ms

Status code distribution:
  [OK]   300000 responses
```

And this is how they look like with always enabled `cancelled` event:

<img width="1728" alt="Screen Shot 2022-10-17 at 15 49 16" src="https://user-images.githubusercontent.com/1713617/196298014-18ffacf4-4632-43d4-a3e7-c93071a85e70.png">

```
ghz --insecure --proto ./src/hello.proto --call helloworld.Greeter.SayHello -d '{"who":"nodeconf"}' -n 300000 --connections 1 --count-errors --timeout 0 0.0.0.0:8080

Summary:
  Count:	300000
  Total:	10.94 s
  Slowest:	23.66 ms
  Fastest:	0.15 ms
  Average:	1.50 ms
  Requests/sec:	27428.62

Response time histogram:
  0.147  [1]      |
  2.498  [281007] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  4.850  [16552]  |∎∎
  7.201  [1660]   |
  9.553  [431]    |
  11.904 [160]    |
  14.256 [32]     |
  16.607 [7]      |
  18.959 [35]     |
  21.311 [15]     |
  23.662 [100]    |

Latency distribution:
  10 % in 0.98 ms
  25 % in 1.16 ms
  50 % in 1.33 ms
  75 % in 1.59 ms
  90 % in 2.24 ms
  95 % in 2.62 ms
  99 % in 4.45 ms

Status code distribution:
  [OK]   300000 responses
```

I understand that this isn't strictly grpc-js making things slow, but being unable to differentiate whether `cancelled` event is an actual ungraceful termination from either end or a graceful one makes it hard to create further conditional logic

Reply:
I think the right approach here, instead of looking for the `aborted` event, is to track whether the stream has finished gracefully and avoid emitting the `cancelled` event if it has. I think the right thing to check for here is whether the server has sent the status (i.e. whether `sendStatus` has been called).

Reply:
definitely not opposed to the idea, though I believe we need to take into account both client and server cancelling the call, this is a little beyond my current understanding of the codebase and would likely require quite a bit more time to ensure nothing breaks, though if you can point me to specific places where such tracking makes sense it would be great!

Reply:
My suggestion is just to add a boolean field (call it `statusSent` or something) to `Http2ServerCallStream`, with the default value `false`, and set it to `true` in `sendStatus`, and change this line to `if (!statusSent)`.

Reply:
Done what we've discussed in the new commit